### PR TITLE
feat: Add token caching for ClientCredentials + DeviceFlow + Auth0 Functionality

### PR DESCRIPTION
**Pull request by PudgyPigeon**
# TL;DR
Adds token caching for ClientCredentials/DeviceFlow and also adds support for Auth0.

## Type
 - [x ] Bug Fix
 - [x ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

One change to note is that this PR adds the ability to manually specify the scopes requested during the FlyteClient authentication process. These scopes (if provided manually), are sourced from the `config.yaml` file and will override the default behaviour for finding scopes. This is in support of Auth0 which manually requires requesting OIDC scopes in the request body - but also the idiosyncrasy wherein Auth0 does not allow custom user APIs to be assigned the OIDC scopes explicitly (This leads to an error during Flytepropeller). Our workaround prevents this behaviour and also keeps the default behaviour intact for other IDPs.

The way the `scopes` parameter was parsed in the `token_client` module was incompatible with Auth0. It is also our opinion that it was parsed incorrectly for other IDPs as well. We have changed it to match the PKCE flow's scopes parameter and things are working. Now, the `refresh_token` is able to be retrieved properly. (No refresh token in response in Device/CC flow before.

Also refactored the ClientCredentials + DeviceFlow to use a `Credentials` dataclass object and to store and retrieve this object by calling `Keyring`. This enables caching for Auth in both flows.

Tested over remote SSH and localhost and can confirm all three flows are working with the suggested changes. 

Though we have had to make sure that Keyring was present in our environments. Absence of Keyring may lead to errors using ClientCredentials + DeviceFlow.

Feedback is welcome.

## Tracking Issue
https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

**Comment by PudgyPigeon** [comment]
@wild-endeavor @kumare3 @eapolinario 

If you have the time to look over this and give feedback, that would be more than welcome. We are quite keen to give back to the community.

Hopefully this PR adds functionality and utility for other end-users of Flyte.

Thanks for the great work as always.

**Review by wild-endeavor — COMMENTED** [review]
do you think we can move the saving of the credentials in keyring to another PR?  I'm honestly not sure that that should happen for the non-pkce flows.  esp for something like clientcredentials - usually this is for automated situations like CI, and I think saving the credentials is actually not preferable.  what was your use-case?

Also when you say "The way the scopes parameter was parsed in the token_client module was incompatible with Auth0" - could you given an example of this please?  Where are the scopes coming from?

Also pushed a formatting change to your PR.

**Comment by PudgyPigeon** [comment]
> do you think we can move the saving of the credentials in keyring to another PR? I'm honestly not sure that that should happen for the non-pkce flows. esp for something like clientcredentials - usually this is for automated situations like CI, and I think saving the credentials is actually not preferable. what was your use-case?
> 
> Also when you say "The way the scopes parameter was parsed in the token_client module was incompatible with Auth0" - could you given an example of this please? Where are the scopes coming from?
> 
> Also pushed a formatting change to your PR.

Sorry for the late reply. I will come back to all your comments soon and make the appropriate changes.

File: setup.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 extras_require = {}
 
-__version__ = "0.0.0+develop"
+__version__ = "1.8.0+image-spec"
```
[the remark this answers is not in the document]
Comment (PudgyPigeon) [reply setup.py:5]: Ok easy enough change will keep note and revert back to 0.0.0

File: flytekit/clients/auth_helper.py [hunk 2]
```diff
@@ -90,7 +90,7 @@ def get_authenticator(cfg: PlatformConfig, cfg_store: ClientConfigStore) -> Auth
         return DeviceCodeAuthenticator(
             endpoint=cfg.endpoint,
             cfg_store=cfg_store,
-            audience=cfg.audience,
```
[the remark this answers is not in the document]
Comment (PudgyPigeon) [reply flytekit/clients/auth_helper.py:93]: We made it so that DeviceCodeAuthenticator sets `self._audience = cfg.audience` within `__init__` in its class definition rather than manually passing it in `auth_helper`

We can revert it back, but we thought it would clean up the code but perhaps it was an un-intuitive move.

File: flytekit/clients/auth/authenticator.py [hunk 3]
```diff
@@ -111,7 +122,9 @@ def _initialize_auth_client(self):
                 endpoint=self._endpoint,
                 redirect_uri=cfg.redirect_uri,
                 client_id=cfg.client_id,
-                scopes=cfg.scopes,
+                audience=cfg.audience,  # Only needed for Auth0 - Taken from client config
+                scopes=self._scopes
+                or cfg.scopes,  # If scope passed in during object instantiation, that takes precedence over cfg.scopes - FOR AUTH0
```
[the remark this answers is not in the document]
Comment (PudgyPigeon) [reply flytekit/clients/auth/authenticator.py:127]: I think the changes should be in this PR but we added a `scopes` field into the `config.yaml` parser.

Basically for PLATFORM/CONFIGURATION classes, we've added a list called scopes as param to be parsed from config.yaml, if that makes sense.

And we set it so it takes precedence over cfg.scopes - and wrote it in this way so as to not break current compatibility. Hopefully that makes sense, I can elaborate more if need be.

**Comment by PudgyPigeon** [comment]
> ?
> 
> Also pushed a formatting change to your PR.

Also in terms of caching - we have a business use case where we wish to cache access/refresh tokens with Auth0. If we don't modify the scopes being passed into the API requests, we can't get the appropriate tokens as a response - therefore no caching. So this was a necessary change to make Keyring work with Auth0.

In regards to ClientCredentials, I agree there's no point in caching a refresh token - but we do have a business directive to limit the amount of requests made to Auth0 - and so we cache the access_token even in the ClientCredentials flow. But we do point our end-users to use DeviceFlow for the most part.

File: flytekit/clients/auth/authenticator.py [hunk 4]
```diff
@@ -87,12 +89,20 @@ def refresh_credentials(self):
 class PKCEAuthenticator(Authenticator):
     """
     This Authenticator encapsulates the entire PKCE flow and automatically opens a browser window for login
+
+    For Auth0 - you will need to manually configure your config.yaml to include a scopes list of the syntax:
```
[the remark this answers is not in the document]
Comment (PudgyPigeon) [reply flytekit/clients/auth/authenticator.py:93]: Really the only scenario that requires this modified scope is in regards to Auth0 and refresh_tokens/caching. Other IDPs probably don't need this change, but the unmodified SDK auth flow resulted in the Auth0 logs showing us that no scopes were ever requested in the auth process.
